virtualcluster/pkg/controller/controllers/capi_virtualcluster_controller.go

```diff
@@ -0,0 +1,169 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/go-logr/logr"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	tenancyv1alpha1 "sigs.k8s.io/cluster-api-provider-nested/virtualcluster/pkg/apis/tenancy/v1alpha1"
+	"sigs.k8s.io/cluster-api-provider-nested/virtualcluster/pkg/controller/constants"
+	"sigs.k8s.io/cluster-api-provider-nested/virtualcluster/pkg/controller/controllers/provisioner"
+	kubeutil "sigs.k8s.io/cluster-api-provider-nested/virtualcluster/pkg/controller/util/kube"
+	strutil "sigs.k8s.io/cluster-api-provider-nested/virtualcluster/pkg/controller/util/strings"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha4"
+)
+
+var (
+	_ reconcile.Reconciler = &ReconcileCAPIVirtualCluster{}
+
+	ownerKey = ".metadata.controller"
+	apiGVStr = tenancyv1alpha1.SchemeGroupVersion.String()
+)
+
+// ReconcileCAPIVirtualCluster reconciles a VirtualCluster object
+type ReconcileCAPIVirtualCluster struct {
+	client.Client
+	Log             logr.Logger
+	ProvisionerName string
+	Provisioner     provisioner.Provisioner
+}
+
+// SetupWithManager will configure the VirtualCluster reconciler
+func (r *ReconcileCAPIVirtualCluster) SetupWithManager(mgr ctrl.Manager, opts controller.Options) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		WithOptions(opts).
+		For(&tenancyv1alpha1.VirtualCluster{}).
+		Owns(&clusterv1.Cluster{}).
```

Comment:
sorta, right now if the cluster isn't up we blindly retry until successful, which accounts for this case, the code I removed wasn't being used at all since I didn't set and ownerref.

Reply:
my concern is I'm not sure I fully agree that a VC should be the owner of the cluster, like there could be a case where you want to move this can not delete all the resources and having an ownerref would setup GC to delete it.

Reply:
This all happens right - https://github.com/kubernetes-sigs/cluster-api-provider-nested/pull/136/files#diff-607056cc549543fc88ebc9695705629f15915ccaac3161e854845ca5a2f0b548R149-R159

We can definitely improve that too trying to figure out the right model for how this wires together.

Reply:
My concern is that if vc is not the cluster owner, once cluster become ready, the event will not trigger vc reconciler because the cluster does not have owner. Then the vc status cannot be updated to ready unless to add periodic check. Do I miss anything?

Reply:
Yeah, check the link in that last message, if the cluster is in Provisioning which is auto set if blank then I `return ctrl.Result{RequeueAfter: 1*time.Second}, nil`

Reply:
I see. Since vc lifecycle and cluster lifecycle are completed decoupled now, polling in VC controller may not be ideal since it may take longtime until the cluster is ready. It is ok for now, moving forward we can set up cluster `Watch` manually and implement a custom enqueue struct so that we can add the matching VC in the queue by checking the annotation.

Reply:
Yeah, let me file an issue to track on improving this, cause I 100% agree.